exec/java-exec/src/test/java/org/apache/drill/test/OperatorTestBuilder.java

```diff
@@ -0,0 +1,300 @@
+            totalBytes.add(columnSize.getTotalDataSize());
+          }
+        }
+
+        for (int columnIndex = 0; columnIndex < numColumns; columnIndex++) {
+          final ValueVector valueVector = actualBatchContainer
+            .getValueVector(columnIndex)
+            .getValueVector();
+
+          if (valueVector instanceof FixedWidthVector) {
+            ((FixedWidthVector) valueVector).allocateNew(actualTotalRows);
+          } else if (valueVector instanceof VariableWidthVector) {
+            final MutableInt totalBytes = totalBytesPerColumn.get(columnIndex);
+            ((VariableWidthVector) valueVector).allocateNew(totalBytes.getValue(), actualTotalRows);
+          } else {
+            throw new UnsupportedOperationException();
+          }
+        }
+
+        try {
+          int currentIndex = 0;
+
+          for (RowSet actualRowSet: actualResults) {
+            final Copier copier;
+            final VectorContainer rowSetContainer = actualRowSet.container();
+            rowSetContainer.setRecordCount(actualRowSet.rowCount());
+
+            switch (actualRowSet.indirectionType()) {
+              case NONE:
+                copier = new GenericCopier();
+                break;
+              default:
+                throw new UnsupportedOperationException();
+            }
+
+            copier.setup(rowSetContainer, actualBatchContainer);
+            copier.appendRecords(currentIndex, actualRowSet.rowCount());
+            currentIndex += actualRowSet.rowCount();
+          }
+        } finally {
+          actualBatch.clear();
+        }
+      } else {
+        // Compare expected and actual results
+        for (int batchIndex = 0; batchIndex < expectedNumBatches; batchIndex++) {
+          final RowSet expectedBatch = expectedResults.get(batchIndex);
+          final RowSet actualBatch = actualResults.get(batchIndex);
+          new RowSetComparison(expectedBatch).verify(actualBatch);
+        }
+      }
+    } finally {
+      // free resources
+
+      if (testOperator != null) {
+        testOperator.close();
+      }
+
+      actualResults.forEach(rowSet -> rowSet.clear());
+
+      if (expectedResults != null) {
+        expectedResults.forEach(rowSet -> rowSet.clear());
+      }
+
+      upstreamBatches.forEach(rowSetBatch -> {
+        try {
+          rowSetBatch.close();
+        } catch (Exception e) {
+          logger.error("Error while closing RowSetBatch", e);
+        }
+      });
+    }
+  }
+
+  /**
+   * Make sure the inputs are valid.
+   */
+  private void validate() {
+    if (combineOutputBatches) {
+      Preconditions.checkArgument(expectedResults.isEmpty() || expectedResults.size() == 1,
+        "The number of expected result batches needs to be zero or one when combining input batches");
+      Preconditions.checkArgument((expectedResults.isEmpty() && (!expectedNumBatchesOpt.isPresent() && expectedTotalRowsOpt.isPresent())) ||
```

Comment:
Fixed defined to define.

The preconditions are different. Specifically these two parts of the conditions.

```
(!expectedNumBatchesOpt.isPresent() && expectedTotalRowsOpt.isPresent())
(expectedNumBatchesOpt.isPresent() || expectedTotalRowsOpt.isPresent())
```

They are different because when you are combining output batches it doesn't make sense to check the number of output batches in the test.

Reply:
makes sense. Thanks for explanation.